Add Local Storage Capacity Isolation API

jingxu97 opened the pull request:
This PR adds the new APIs to support storage capacity isolation as
described in the proposal [https://github.com/kubernetes/community/pull/306](url)

1. Add SizeLimit for emptyDir volume
2. Add scratch and overlay storage type used by container level or
node level


**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
Alpha feature: Local volume Storage Capacity Isolation allows users to set storage limit to isolate EmptyDir volumes, container storage overlay, and also supports allocatable storage for shared root file system. 
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44785)
<!-- Reviewable:end -->

nikhiljindal commented:
Are you sure you wanted to refer https://github.com/kubernetes/kubernetes/pull/306?

Someone from @kubernetes/sig-storage-api-reviews should be able to review this better.
Whoever wants to take this, please feel free to remove me as reviewer.

vishh commented on pkg/api/types.go:
```diff
@@ -2944,6 +2949,10 @@ const (
 	// Volume size, in bytes (e,g. 5Gi = 5GiB = 5 * 1024 * 1024 * 1024)
 	ResourceStorage ResourceName = "storage"
+	// Local Storage for overlay filesystem, in bytes. (500Gi = 500GiB = 500 * 1024 * 1024 * 1024)
+	ResourceStorageOverlay ResourceName = "storage.kubernetes.io/overlay"
```
Clarify that the resource names are alpha and that they can change across releases. This is important because a new concept of `ResourceClasses` is being discussed where not having a domain name in the user resource name request will be useful.

ddysher replied:
Apart from the domain part, it might be necessary to clarify 'overlay' to make sure people won't confused about the "overlay filesystem driver" and general overlay mechanism used in container, which I think we mean the later here.

derekwaynecarr replied:
@vishh -- why not just name it alpha like gpu?

derekwaynecarr replied:
i guess it doesnt matter if its behind the feature gate really.

vishh replied:
That will require renaming configs post alpha.

jingxu97 replied:
fixed

vishh commented on pkg/api/validation/validation.go:
```diff
@@ -383,7 +383,12 @@ func validateVolumeSource(source *api.VolumeSource, fldPath *field.Path) field.E
 		numVolumes++
-		// EmptyDirs have nothing to validate
+		if !utilfeature.DefaultFeatureGate.Enabled(features.LocalStorageCapacityIsolation) {
+			unsetSizeLimit := resource.Quantity{}
```
How about validating Pod Spec resources?

jingxu97 replied:
add validation for ResourceRequirements which will be called by validateContainer

derekwaynecarr commented on pkg/api/types.go:
```diff
@@ -580,6 +580,11 @@ type EmptyDirVolumeSource struct {
 	// +optional
 	Medium StorageMedium
+	// Total amount of local storage required for this directory.
+	// The default is nil which means that the directory can use all available local storage on the node.
```
its not always all available local storage on the node.  maybe just say its undefined if omitted.

thockin replied:
agree

derekwaynecarr commented on pkg/api/validation/validation.go:
```diff
@@ -383,7 +383,12 @@ func validateVolumeSource(source *api.VolumeSource, fldPath *field.Path) field.E
+		if !utilfeature.DefaultFeatureGate.Enabled(features.LocalStorageCapacityIsolation) {
+			unsetSizeLimit := resource.Quantity{}
+			if unsetSizeLimit.Cmp(source.EmptyDir.SizeLimit) != 0 {
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("emptyDir").Child("sizeLimit"), "LocalStorageCapacityIsolation are disabled by feature-gate"))
```
nit: is disabled

jingxu97 replied:
fixed

thockin commented:
/approve

jingxu97 committed:
Add storage isolation API

This PR adds the new APIs to support storage capacity isolation as described in the proposal
https://github.com/kubernetes/community/pull/306

1. Add SizeLimit for emptyDir volume
2. Add scratch and overlay storage type used by container level or
node level

jingxu97 commented:
@k8s-bot pull-kubernetes-node-e2e test this

jingxu97 commented:
@thockin @vishh @derekwaynecarr PTAL

jingxu97 committed:
generated files

jingxu97 commented:
Rebased. @thockin @vishh @derekwaynecarr PTAL

derekwaynecarr commented on pkg/api/types.go:
```diff
@@ -601,6 +601,12 @@ type EmptyDirVolumeSource struct {
+	// The default is nil which means that the limit is undefined.
+	// More info: http://kubernetes.io/docs/user-guide/volumes#emptydir
+	// +optional
+	SizeLimit resource.Quantity
```
I want to understand how this field will work with memory backed volumes.

For example, if a pod has a memory request for 500Mi of memory, and has a memory backed volume of 1Gi, the total memory request is still 500Mi, correct?  Users are responsible for ensuring that memory backed volumes are sized as part of their pod requests similar to how /dev/shm works?  In addition, a BestEffort pod can have a memory backed volume that has a size limit without changing the pod qos?

jingxu97 replied:
Yes, the total memory request is still 500Mi. When specifying the size limit for emptydir, it does not change pod qos

derekwaynecarr commented:
Assuming the size limit for memory backed volumes does not impact QoS or how the pod cgroups are managed, this is LGTM.  The scheduler still schedules based on resource requirements only.

derekwaynecarr commented:
/lgtm

vishh commented on pkg/api/validation/validation.go:
```diff
@@ -396,7 +396,12 @@ func validateVolumeSource(source *api.VolumeSource, fldPath *field.Path) field.E
+		if !utilfeature.DefaultFeatureGate.Enabled(features.LocalStorageCapacityIsolation) {
+			unsetSizeLimit := resource.Quantity{}
+			if unsetSizeLimit.Cmp(source.EmptyDir.SizeLimit) != 0 {
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("emptyDir").Child("sizeLimit"), "LocalStorageCapacityIsolation is disabled by feature-gate"))
```
nit: Make the error user friendly - `SizeLimit field disabled by feature-gate for EmptyDir volumes`

jingxu97 replied:
fixed

vishh reviewed (approved):
Just one nit, otherwise LGTM

vishh commented:
I removed the LGTM label to fix the pending nit and add a note on how EmptyDir SizeLimits apply to memory backed volumes as discussed offline.

jingxu97 commented:
@vishh updated the comment, PTAL

jingxu97 commented:
updated the comments, add /lgtm back

vishh commented:
/lgtm

jingxu97 commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

jingxu97 commented:
bump priority to make sure API change for this feature gets in
